services/m3dbnode/config/config.go

```diff
@@ -190,3 +205,146 @@ type HashingConfiguration struct {
 	// Murmur32 seed value.
 	Seed uint32 `yaml:"seed"`
 }
+
+// ETCDConfig creates a new embedded etcd config from kv config.
+func ETCDConfig(cfg Configuration) (*embed.Config, error) {
+	newKVCfg := embed.NewConfig()
+	kvCfg := cfg.EnvironmentConfig.EmbeddedServer
+
+	dir := kvCfg.Dir
+	if dir == "" {
+		dir = path.Join(cfg.Filesystem.FilePathPrefix, defaultETCDDirSuffix)
+	}
+	newKVCfg.Dir = dir
+
+	// listen-peer-urls
+	LPUrls, err := convertToURLsWithDefault(kvCfg.LPUrls, newURL(defaultETCDListenHost, defaultETCDServerPort))
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.LPUrls = LPUrls
+
+	// listen-client-urls
+	LCUrls, err := convertToURLsWithDefault(kvCfg.LCUrls, newURL(defaultETCDListenHost, defaultETCDClientPort))
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.LCUrls = LCUrls
+
+	host, err := getHostFromHostID(kvCfg.InitialCluster, kvCfg.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// initial-advertise-peer-urls
+	APUrls, err := convertToURLsWithDefault(kvCfg.APUrls, newURL(host, defaultETCDServerPort))
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.APUrls = APUrls
+
+	// advertise-client-urls
+	ACUrls, err := convertToURLsWithDefault(kvCfg.ACUrls, newURL(host, defaultETCDClientPort))
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.ACUrls = ACUrls
+
+	newKVCfg.Name = kvCfg.Name
+	newKVCfg.InitialCluster = kvCfg.InitialCluster
+
+	copySecurityDetails := func(tls *transport.TLSInfo, ysc *environment.KVSecurity) {
+		tls.CAFile = ysc.CAFile
+		tls.CertFile = ysc.CertFile
+		tls.KeyFile = ysc.KeyFile
+		tls.ClientCertAuth = ysc.CertAuth
+		tls.TrustedCAFile = ysc.TrustedCAFile
+	}
+	copySecurityDetails(&newKVCfg.ClientTLSInfo, &kvCfg.ClientSecurityJSON)
+	copySecurityDetails(&newKVCfg.PeerTLSInfo, &kvCfg.PeerSecurityJSON)
+	newKVCfg.ClientAutoTLS = kvCfg.ClientSecurityJSON.AutoTLS
+	newKVCfg.PeerAutoTLS = kvCfg.PeerSecurityJSON.AutoTLS
+
+	return newKVCfg, nil
+}
+
+func newURL(host string, port int) string {
+	return fmt.Sprintf("%s:%d", host, port)
+}
+
+func convertToURLsWithDefault(urlStrs []string, def ...string) ([]url.URL, error) {
+	if len(urlStrs) == 0 {
+		urlStrs = def
+	}
+
+	urls, err := types.NewURLs(urlStrs)
+	if err != nil {
+		return nil, err
+	}
+
+	return []url.URL(urls), nil
+}
+
+func getHostFromHostID(initialCluster, hostID string) (string, error) {
+	if len(initialCluster) == 0 {
+		return "", errors.New("zero nodes in initialCluster")
+	}
+
+	nodesStr := strings.Split(initialCluster, ",")
+
+	for _, nodeStr := range nodesStr {
+		nodeData := strings.Split(nodeStr, "=")
+
+		if len(nodeData) != 2 {
+			return "", errors.New("invalid initialCluster format")
+		}
+
+		if hostID == nodeData[0] {
```

Comment:
As discussed at some point the hostID may not be the same as the IP:port listed in the seed nodes list.

Can we instead add the concept of every seed host with: